src/ring_theory/tensor_product.lean

```diff
@@ -1,43 +1,247 @@
 /-
 Copyright (c) 2020 Scott Morrison. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Scott Morrison
+Authors: Scott Morrison, Johan Commelin
 -/
 
 import linear_algebra.tensor_product
-import algebra.algebra.basic
+import algebra.algebra.tower
 
 /-!
 # The tensor product of R-algebras
 
-We construct the R-algebra structure on `A ⊗[R] B`, when `A` and `B` are both `R`-algebras,
-and provide the structure isomorphisms
+Let `R` be a (semi)ring and `A` an `R`-algebra.
+In this file we:
 
-* `R ⊗[R] A ≃ₐ[R] A`
-* `A ⊗[R] R ≃ₐ[R] A`
-* `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+- Define the `A`-module structure on `A ⊗ M`, for an `R`-module `M`.
+- Define the `R`-algebra structure on `A ⊗ B`, for another `R`-algebra `B`.
+  and provide the structure isomorphisms
+  * `R ⊗[R] A ≃ₐ[R] A`
+  * `A ⊗[R] R ≃ₐ[R] A`
+  * `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
+
+## Main declaration
+
+- `linear_map.base_change A f` is the `A`-linear map `A ⊗ f`, for an `R`-linear map `f`.
 
-The code for
-* `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
-is written and compiles, but takes longer than the `-T100000` time limit,
-so is currently commented out.
 -/
 
 universes u v₁ v₂ v₃ v₄
 
-namespace algebra
-
 open_locale tensor_product
 open tensor_product
 
 namespace tensor_product
 
+variables {R A M N P : Type*}
+
+/-!
+### The `A`-module structure on `A ⊗[R] M`
+-/
+
+open linear_map
+open algebra (lsmul)
+
+namespace algebra_tensor_module
+
+section semiring
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [module R M] [module A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [module R N]
+variables [add_comm_monoid P] [module R P] [module A P] [is_scalar_tower R A P]
+
+lemma smul_eq_lsmul_rtensor (a : A) (x : M ⊗[R] N) : a • x = (lsmul R M a).rtensor N x := rfl
+
+/-- Heterobasic version of `tensor_product.curry`:
```

Comment:
But maybe that's not interesting since it still requires `A` to be a semiring

Reply:
I agree this is probably the right statement. Indeed this could go in `linear_algebra/tensor_product.lean` and, I think, actually replace `tensor_product.curry` (via `smul_comm_class R R M` etc). The same goes for some other defs here.

I've shied away from aiming for greater generality here and elsewhere because I'm shooting to unblock extension of scalars and if I really step back and think about where we should eventually get, I see [quite a lot of work](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.234773.20base.20change/near/240929258).

What do you think? I'm certainly happ just to adjust this def as you suggest, see if I can thus replace `tensor_product.curry`, and leave it at that.

Reply:
I think generalizing away `algebra` probably isn't all that useful here; the interesting generalizations I can think of require `distrib_mul_action_hom` instead of `linear_map`, and we don't have the theory of those developed (even though in all cases where both exist they're isomorphic).

> Indeed this could go in linear_algebra/tensor_product.lean and, I think, actually replace tensor_product.curry

This is quite tempting, but probably introduces some extra type arguments that would be annoying to pass explicitly. If we went down this route, I can imagine wanting `curry'` as your new def and `curry` as the less-complexly-typed version implemented with `curry`.

Perhaps the thing to do for now is add a comment somewhere in this file saying something along the lines of
```lean
/- As these heterobasic definitions are just more general versions of the definitions already in
`linear_algebra/tensor_product`, we could consider replacing the less general definitions with these ones.
If we do this, we probably should still implement the less general ones as abbreviations to the more general
ones with fewer type arguments. -/
````

Reply:
Agreed, and done.